Add MTESolarFactory

purebluez opened the pull request:
Update/Continuation of https://github.com/GTNewHorizons/GT5-Unofficial/pull/2281

Related pr for recipes: https://github.com/GTNewHorizons/NewHorizonsCoreMod/pull/1120

NOTE: This multi is NOT meant to make Solar Panels as power generation better/more viable than they are now. It is only meant to improve user experience when automating them. If needed, I will nerf the recipes as they are currently 1:1 with dire table.

Additional notes of importance:
- This PR does not affect the balance of Compressed Solar Panels, as it cannot craft the 1V panels that Compressed Solar Panels need.


- The multi is not a MuTE like the original pr said it would come back as. This is due to the MuTE system being officially dead.

Basic idea of the multi:
---
The Solar Factory acts as a GT Replacement for the tiered Solar Panel recipes which are currently made in the Dire (Auto)Crafting Table. Since these Solar Panels are used heavily in some endgame recipes, the Dire Tables quickly become a bottleneck as well as a major annoyance since you have to manually set the recipe one item at a time if you want to automate them.

Currently the goal of the multi is not to make the Solar Panels easy to craft for power generation purposes.

Complaints I have addressed from the original pr (see above):
--- 

- Steel Casings are not appropriate for a ZPM multiblock

I have changed the structure to use higher tier casings. I think they look OK and are OK balance-wise, but I would like to hear if people think they should be changed to a different casing, or if I should make a whole new casing for the multi. 
Below there is a screenshot of the structure in full, with the altered casings.

- The multi is boring / has no mechanic.

I have given it a new mechanic. Below there is a screenshot of the tooltip which explains it in detail, but the gist is that every solar has a minimum wafer tier in the recipe and using a higher tier wafer multiplies how many solar panels you get per recipe **after parallels.**

Speaking of parallels, There are now Precise Unit Casings on the T2 and T3 structures. These work the same way as they do with the Precise Assembler, but with 1/4th the parallels and no Imprecise Unit Casing allowed since I felt T1 Precise was more in line with the tiering of the multi. The parallels are shown in the tooltip as well but are as follows:
MK-I = 8x, MK-II = 16x, MK-III = 32x, MK-IV = 64x

A couple examples:
LV panel recipe uses T2 (Phosphorus doped) wafers as a minimum. If you use T4 (Europium doped) wafers, the recipe will output 50% more LV panels. This means with 2x consumed inputs, you would get 3x panels.

ZPM panel recipe uses 8 T5 (Americium doped) wafers as a minimum. If you use 8 T4 wafers or 6 T5 wafers, the recipe wont run as wafer used is lower than the minimum tier for the former, and the wafer count is too low for the latter. If you wanted to use higher tier wafers for more output on this recipe, you would need to use either Photonically Prepared (T6) or Photonically Enhanced (T7).

- The NEI page for the recipes is buggy and looks bad.

I removed the custom background icon and the category icon is now normal. This one is more of a bug fix but there was a complaint about it.

Changes from the original that weren't complained about:
---
- Removed solar cells, 

Now you can directly craft into Panels (UV/UHV+) or follow a similar system to current dire tables where every panel requires the previous (HV+)

- Downtiered multi to HV, from ZPM.

- Gave the multi 2 new tiers

This one was not complained about, but instead a suggestion. For more information on the structure and processing changes the new tiers involve, look below to where I have the image of the tooltip and all 3 structures.

Structure and Tooltip:
---
From left to right, these are the structures for the T1, T2, and T3 versions of the Solar Factory.
![image](https://github.com/user-attachments/assets/3fdfeb64-f2c7-490a-a502-c47a768aaa12)
![solarfactorytooltip](https://github.com/user-attachments/assets/4074f59e-9e60-4dae-9f8e-095efbd29a1d)


For the tooltip, I felt there should be information on what wafers count and what tier they are, but a friend believes it would be bloat and what wafers the tooltip are talking about is obvious. Please give feedback on if its unclear or not.

Recipes:
---

For the controller, meant to be about HV. It is any EV circuit, not specifically nano assembly
![image](https://github.com/user-attachments/assets/ccc49cc5-2576-4929-a45c-f8fd49f8e329)

The Recursive Solar Panel recipes are 1:1 replicas of the dire table recipes, but with the following minor changes:

Plastic plates (epoxy, ptfe, pbi, etc.) in all recipes are now molten with the equivalent amount (144L per plate)
IV solar recipe uses 8 T3 wafers instead of 4 T3 and 4 T2
LuV solar recipe uses 8 T4 wafers instead of 4 T4 and 4 T3
ZPM solar recipe uses 8 T5 wafers instead of 4 T5 and 4 T3
ZPM and UV solar recipes no longer uses any carbon plates (previously 4)
UV solar recipe no longer uses any Solar Grade Silicon 

Please give feedback on if they should be made more expensive or less expensive.
|| Recursive | Non-Recursive |
| ------------- | ------------- | ------------- |
|LV|![LV1](https://github.com/user-attachments/assets/a2fba6fd-d5d9-4fbd-b984-1441016d1663) | ![LV2](https://github.com/user-attachments/assets/851cea41-ca8a-4685-883a-8bf10679a591) |
|MV|![MV1](https://github.com/user-attachments/assets/d0986e90-5dab-4ef7-a72e-d0a0c095bb37) |![MV2](https://github.com/user-attachments/assets/be298fe1-10c1-4a5f-806c-215539a9135e) |
|HV|![HV1](https://github.com/user-attachments/assets/15565bda-5269-468d-9c8a-c86cc24b418f) |![HV2](https://github.com/user-attachments/assets/c093481f-1f5e-45f5-80ae-f9ffa4cfd468) |
|EV|![EV1](https://github.com/user-attachments/assets/df56a101-b1c8-4585-a95a-4d420cdf38c5)  |![EV2](https://github.com/user-attachments/assets/1d947b4a-fc06-4ac0-a11c-62e9e3f9733a) |
|IV|![IV1](https://github.com/user-attachments/assets/df29e762-89a5-4ec2-85af-fc830bcf15d9) | ![IV2](https://github.com/user-attachments/assets/ba2b046f-ebd6-40a7-861b-c9a929f36278) |
|LuV| ![LuV1](https://github.com/user-attachments/assets/5845d793-4706-4dcf-ba6b-5da5dee5f698) | ![LuV2](https://github.com/user-attachments/assets/cf21f73c-9763-4e47-94b6-a39426273f5d)|
|ZPM|![ZPM1](https://github.com/user-attachments/assets/df849602-1aa7-4149-8d18-424abe551d4f) | ![ZPM2](https://github.com/user-attachments/assets/522289e7-57e6-4686-b3e1-531fddf6a9fa)|
|UV| ![UV1](https://github.com/user-attachments/assets/2bef2102-6da7-4846-a050-a2ecca14d6f9) | ![UV2](https://github.com/user-attachments/assets/eb30a66a-e52b-407f-bc6b-eff6466f611e)|

Closes https://github.com/GTNewHorizons/areas-of-interest/issues/6

FourIsTheNumber commented on src/main/java/gregtech/common/tileentities/machines/multi/MTESolarFactory.java:
```diff
@@ -0,0 +1,397 @@
+            .addInfo("The recipes shown in NEI display the minimum wafer tier required")
+            .addInfo("Parallels are based on Precise Casing Tier")
+            .addInfo("MK-I = 8x, MK-II = 16x, MK-III = 32x, MK-IV = 64x")
+            .addInfo("Supports " + TT + " energy hatches")
```
We have a standard .addTecTechHatchInfo() now

EnderProyects commented:
I think that the recipes of the solars should include the previous one as in the dire recipes, because if not there is no reason to have all the recipes, people are just going to use the last one

EnderProyects commented:
If you need more space just add 3 more slots to the machine ( and could add as well 2 more for fluids) so if in the future there is some upgrade to solars there is space for recipes

chochem commented:
Have you brought this to the admins before making the PR?

Please note the new procedures if you missed them https://discord.com/channels/181078474394566657/1243216765727539383/1317949621510733936

Dream-Master commented:
I remember there was such solar tower suggestion before pr . It’s been a while . The idea is nice but a bit to simple . Maybe add tiers like the pbc or so. 
but chochom is right start a conversation first in areas of interest . Ask us to open a thread and things can be discussed. If you code it first there is a possibility to reject the project and all your work is done for nothing .

Dream-Master commented:
@purebluez https://discord.com/channels/181078474394566657/1321055049324953620

serenibyss commented on src/main/java/gregtech/api/util/GTRecipeConstants.java:
```diff
@@ -201,6 +202,12 @@ public class GTRecipeConstants {
+    /**
+     * Solar Factory recipe data containing minimum wafer tier and count
+     */
+    public static final RecipeMetadataKey<SolarFactoryRecipeData> SOLAR_FACTORY_WAFER_DATA = SimpleRecipeMetadataKey
```
I suggest renaming either this or the string to keep the names the same

serenibyss commented on src/main/java/gregtech/loaders/postload/recipes/SolarFactoryRecipes.java:
```diff
@@ -0,0 +1,287 @@
+// Metadata is only required on the hidden recipes.
+
+// If the recipe you are making doesn't use a wafer, just make it without any metadata.
+public class SolarFactoryRecipes implements Runnable {
```
For testing these are fine, though before merge these should probably be moved to coremod to remove the getModItem() call on coremod items

serenibyss commented on src/main/java/gregtech/common/tileentities/machines/multi/MTESolarFactory.java:
```diff
@@ -0,0 +1,395 @@
+
+    @NotNull
+    @Override
+    public CheckRecipeResult checkProcessing() {
```
It would be best if this override could be avoided and instead move custom logic into ProcessingLogic overrides

serenibyss commented on src/main/java/gregtech/common/tileentities/machines/multi/MTESolarFactory.java:
```diff
@@ -0,0 +1,395 @@
+        return new ITexture[] { Textures.BlockIcons.getCasingTextureForId(CASING_INDEX) };
+    }
+
+    @Override
+    public int getMaxEfficiency(ItemStack aStack) {
+        return 10000;
+    }
+
+    @Override
+    public boolean isCorrectMachinePart(ItemStack aStack) {
+        return true;
+    }
+
+    @Override
+    public int getDamageToComponent(ItemStack aStack) {
+        return 0;
+    }
+
+    @Override
+    public boolean explodesOnComponentBreak(ItemStack aStack) {
+        return false;
+    }
+
+    @Override
+    public boolean supportsBatchMode() {
+        return true;
+    }
+
+    @Override
+    public boolean supportsVoidProtection() {
+        return true;
+    }
+
+    @Override
+    public boolean supportsInputSeparation() {
+        return true;
+    }
```
Many of these overrides are probably not needed, since they may be identical to the super method

serenibyss commented on src/main/java/gregtech/api/util/GTLanguageManager.java:
```diff
@@ -550,6 +550,7 @@ public static void writePlaceholderStrings() {
         addStringLocalization("Interaction_DESCRIPTION_Index_507", "Safe Mode");
         addStringLocalization("Interaction_DESCRIPTION_Index_508", "Requires Stabilized Black Hole");
         addStringLocalization("Interaction_DESCRIPTION_Index_509", "Requires HIP Unit");
+        addStringLocalization("Interaction_DESCRIPTION_Index_510", "Minimum wafer tier: ");
```
This doesn't seem to be used

purebluez commented:
I've received a few suggestions and seen some feedback and will be implementing them soon enough. Marking this as draft until I feel its ready to merge again.

Dream-Master commented:
> I've received a few suggestions and seen some feedback and will be implementing them soon enough. Marking this as draft until I feel its ready to merge again.

Did it has different tiers ?  Like pcb multi ?

purebluez commented:
> Did it has different tiers ?  Like pcb multi ?

Im planning on 3 tiers, and adding back the recursive solar recipes. Here is a tentative plan on the tiers: 

T1:
- HV
- Can only do recursive recipes
- No wafer output multiplication

T2:
- LuV (same as now) 
- Can only do recursive recipes
- Wafer output multiplication allowed

T3:
- UHV
- Can do non-recursive recipes 
- Wafer output multiplication allowed

This might be a little basic, but I'm not sure what else should upgrade every tier. Do you have any thoughts?

purebluez commented:
I have changed the current recipes to have the previous tiered solar panel, just like how they are currently. I have also added non-recursive recipes that are gated later than previously and made much more expensive, once I make the 2 other planned tiers of the multi the non-recursive recipes will be gated behind the final tier, likely gated at UHV or UEV. 

PR Description updated with the new recipes.

EnderProyects commented:
I like the idea that the advanced recipes uses the advanced Irradiant plate... but you are removing all the magic materials that it used in the other recipe plates...
![image](https://github.com/user-attachments/assets/5f87b1e6-7074-4156-a2ee-312ea6ac9e39)

purebluez commented:
> I like the idea that the advanced recipes uses the advanced Irradiant plate... but you are removing all the magic materials that it used in the other recipe plates... 

I took this into account. for example:
LuV solar is used alot in dyson, so i added fiery steel, knightmetal, and stealeaf to it. Other solars that normally have a tier of irradiant plate that uses magic ingredients have had molten versions of that material added to the Solar Factory non-recursive recipe for that panel.

I think UV solar is used in some EoH parts as well, but I didnt add the magic ingredients to that recipe since by UMV wouldnt it be silly to need to automate these twilight forest materials?

EnderProyects commented:
Kinda you now need the TF materials for EoH so is a big buff not neding them now, I would say to add molten knightmetal and dense stealeaf to the two slots that there are in the recipe

purebluez commented:
I have added the new structures for the multi and updated PR description accordingly. Will un-draft now that the new tiers are done.

purebluez commented:
> I would say to add molten knightmetal and dense stealeaf to the two slots that there are in the recipe

Done. I have also updated the PR description recipe table section to reflect the change

Dream-Master commented:
We can test it in nightly ?

purebluez commented:
> We can test it in nightly ?

Yes, it should work fine.

Dream-Master commented:
@purebluez so this recipes will replace the 9x9 crasfts right ?

purebluez commented:
> @purebluez so this recipes will replace the 9x9 crasfts right ?
> 

Not right now since it doesn't remove the recipes, but there wouldn't be significant balance changes if they did.

Dream-Master commented:
ok i compare recipes later

serenibyss reviewed (approved):
Code looks good, and we decided this has been in testing for long enough now. If anyone has any feedback on the balance of the new recipes, please open a new ticket on the modpack repo
